segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableEntryIterator.java

```diff
@@ -0,0 +1,328 @@
+ */
+package io.pravega.segmentstore.server.tables;
+
+import io.pravega.common.TimeoutTimer;
+import io.pravega.common.util.AsyncIterator;
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.segmentstore.contracts.ReadResult;
+import io.pravega.segmentstore.contracts.tables.TableEntry;
+import io.pravega.segmentstore.server.DirectSegmentAccess;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import io.pravega.segmentstore.server.reading.AsyncReadResultProcessor;
+import lombok.AccessLevel;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Iterates through {@link TableBucket}s in a Segment.
+ * @param <T> Type of the final, converted result.
+ */
+@Slf4j
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@ThreadSafe
+class TableEntryIterator<T> implements AsyncIterator<T> {
+    //region Members
+
+    // The maximum size (in bytes) of each read to perform on the segment.
+    private static final int MAX_READ_SIZE = 2 * 1024 * 1024;
+
+    private final DirectSegmentAccess segment;
+    private final long startOffset;
+    private final int maxLength;
+    private final boolean shouldClear;
+    private final Duration fetchTimeout;
+    private final EntrySerializer entrySerializer;
+    private final ConvertResult<T> resultConverter;
+    private final Executor executor;
+    @GuardedBy("this")
+    private Iterator<Map.Entry<EntryIteratorState, TableEntry>> currentEntry = null;
+    @GuardedBy("this")
+    private long currentBatchOffset;
+    //endregion
+
+    private TableEntryIterator<T> setCurrentBatchOffset(long currentBatchOffset) {
+        this.currentBatchOffset = currentBatchOffset;
+        return this;
+    }
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<T> getNext() {
+        // Verify no other call to getNext() is currently executing.
+        return getNextEntry()
+                .thenCompose(entry -> {
+                    if (entry == null) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return this.resultConverter.apply(entry);
+                    }
+                });
+    }
+
+    public boolean endOfSegment() {
+        return this.currentBatchOffset >= (startOffset + maxLength);
```

Comment:
Nit: either use `this.` for all private fields or for none. Try to be consistent.

Reply:
`this.currentBatchOffset` is Guarded By `this` but you don't use a synchronized block here.